### PR TITLE
shapetracker.py: better view/shapetracker caching

### PR DESCRIPTION
**Pull request by roelofvandijk**
Make most View and Shapetracker methods static. These are mainly pushing tuples around, this should help with caching.
View is now a NamedTuple.

Ultimately ShapeTracker could become a similar Tuple, and each op creates a new one.

Still needs a lot of work on the tests, I'd like to hear if you think it's worth the effort @rayanht @geohot.

`test_net_speed.py`
Master:
```
27.90 lazy.py
15.84 tensor.py
27.06 shapetracker.py
15.83 ops.py
20.01 symbolic.py
9.06 helpers.py
1.07 lib.py
116.76 total
```

This branch:
```
26.30 lazy.py
13.51 tensor.py
16.10 shapetracker.py
13.09 ops.py
1.78 symbolic.py
6.87 helpers.py
0.60 lib.py
78.24 total
```

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/linearizer.py            11              11
tinygrad/lazy.py                           1               1
tinygrad/shape/shapetracker.py           167             120
------------------------------------------------------------
total                                    179             132
------------------------------------------------------------
lines added in the tinygrad folder: 47
```

File: tinygrad/codegen/linearizer.py [hunk 1]
```diff
@@ -169,7 +169,7 @@ def _group_float4(self, i, store_offset):
     return store_offset_float4
 
   def global_load(self, i, idxs:List[Variable], const=None) -> List[Token]:
-    load_offset: Dict[Tuple[int, ...], Any] = {uidxs:(dtypes.float,uidxs)+self.sts[i].expr_idxs(idxs+[Variable.num(x) for x in uidxs[::-1]]) for uidxs in self.shape_offsets(i)}
+    load_offset: Dict[Tuple[int, ...], Any] = {uidxs:(dtypes.float,uidxs)+ShapeTracker.expr_idxs(self.sts[i].views, idxs+[Variable.num(x) for x in uidxs[::-1]]) for uidxs in self.shape_offsets(i)}
```
Comment (guilhermicas) [inline tinygrad/codegen/linearizer.py:172]: `reversed(uidxs)`

File: tinygrad/shape/shapetracker.py [hunk 2]
```diff
@@ -31,70 +34,87 @@ def filter_strides(shape:Tuple[int, ...], strides:Tuple[int, ...]) -> Tuple[int,
...
 def strides_for_shape(shape:Tuple[int, ...]) -> Tuple[int, ...]:
   strides = [1] if shape else []
   for d in shape[::-1][:-1]: strides = [d*strides[0]] + strides
   return tuple([st if s != 1 else 0 for st, s in zip(strides, shape)])
 
-@functools.lru_cache(maxsize=None)
 def view_from_shape(shape:Tuple[int, ...]) -> View:
-  assert all(isinstance(x, int) for x in shape)
-  return View(tuple(shape), strides_for_shape(shape))
+  assert all([isinstance(x, int) for x in shape])
```
Comment (guilhermicas) [inline tinygrad/shape/shapetracker.py:109]: `x.__class__ is int`, there are more in this PR which could be substituted

**Comment by roelofvandijk** [comment]
master (52b7105f8798d88feed0197691391f7b9aa4b013)
```
using <class 'tinygrad.runtime.ops_gpu.CLCodegen'>
testing llama python run time
built model
assigned empty tensors, doing warmup
codegen        runtime (median):  211.05ms, runs:   291.58,  210.03,  207.10,  205.42,  213.85,  255.18,  208.70,  205.04,  212.07,  220.11
methodcache    runtime (median):  216.98ms, runs:   219.24,  268.04,  210.51,  204.78,  244.21,  227.40,  204.73,  243.30,  212.41,  214.72
profile        runtime (median): 1106.56ms, runs:  1291.20, 1067.84, 1067.99, 1124.23, 1129.57, 1105.92, 1086.90, 1107.20, 1192.22, 1091.59
```


this branch
```
using <class 'tinygrad.runtime.ops_gpu.CLCodegen'>
testing llama python run time
built model
assigned empty tensors, doing warmup
codegen        runtime (median):  209.61ms, runs:   273.33,  204.58,  208.70,  194.39,  238.57,  210.01,  206.48,  228.19,  273.12,  209.20
methodcache    runtime (median):  204.24ms, runs:   195.37,  206.58,  246.79,  245.70,  237.84,  198.21,  201.90,  192.73,  194.53,  231.76
profile        runtime (median): 1010.56ms, runs:  1119.41, 1079.45, 1007.60, 1039.65, 1010.28,  997.83, 1010.84, 1001.20, 1116.74, 1004.89
```

**Comment by geohot** [comment]
How does this stuff stack on what's merged?

**Comment by roelofvandijk** [comment]
Perfectly complementary, designed as such. 

Should give another easy 10% or so. I still feel that there's a layer of abstraction that could be removed here, since most ops modify only the last view. I feel that the whole shapetracker layer could be removed, but I need to think a bit more on this. `movement_op` could go as well.

Master (after @rayanht's work, 23648538fa2113c554c3c7589ca1d9b38aeff545):
```
codegen         mean runtime:  144.85ms, runs:   156.84,  127.40,  160.38,  129.25,  124.37,  126.40,  145.71,  156.89,  145.51,  175.79
methodcache     mean runtime:  145.06ms, runs:   126.89,  130.24,  130.69,  129.48,  141.06,  141.02,  139.69,  143.25,  220.84,  147.46
profile         mean runtime:  771.06ms, runs:   954.99,  757.93,  733.25,  755.19,  799.56,  750.36,  726.80,  738.02,  753.20,  741.32
```

This branch:
```
codegen         mean runtime:  137.03ms, runs:   157.13,  160.46,  121.59,  121.39,  125.52,  126.65,  173.97,  126.17,  132.45,  124.97
methodcache     mean runtime:  126.99ms, runs:   122.35,  164.48,  114.08,  121.26,  120.45,  119.81,  118.26,  116.95,  155.76,  116.52
profile         mean runtime:  695.42ms, runs:   742.53,  657.71,  689.83,  730.03,  662.94,  699.96,  689.63,  693.22,  695.22,  693.16
``` 

Full timings:
Master:
```
codegen         mean runtime:  144.85ms, runs:   156.84,  127.40,  160.38,  129.25,  124.37,  126.40,  145.71,  156.89,  145.51,  175.79
methodcache     mean runtime:  145.06ms, runs:   126.89,  130.24,  130.69,  129.48,  141.06,  141.02,  139.69,  143.25,  220.84,  147.46
profile         mean runtime:  771.06ms, runs:   954.99,  757.93,  733.25,  755.19,  799.56,  750.36,  726.80,  738.02,  753.20,  741.32
         6329568 function calls (5883806 primitive calls) in 7681.037 seconds

   Ordered by: internal time
   List reduced from 509 to 51 due to restriction <0.1>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    58632  291.888    0.005 3903.865    0.067 tensor.py:24(apply)
53450/2260  275.770    0.005 2482.475    1.098 lazy.py:109(realize)
    71256  241.126    0.003  590.743    0.008 lazy.py:84(__init__)
    76366  223.503    0.003 1528.787    0.020 lazy.py:68(create_lazybuffer)
   106940  205.651    0.002  246.353    0.002 helpers.py:133(add)
465444/402248  195.375    0.000  244.963    0.001 {built-in method builtins.len}
48096/45078  178.026    0.004 1385.554    0.031 lazy.py:169(shuffle_and_prune_movement_ops)
300152/184332  173.480    0.001  268.296    0.001 {built-in method builtins.hash}
    15086  167.754    0.011 2625.518    0.174 tensor.py:532(_broadcasted)
   118324  154.284    0.001  331.245    0.003 ops.py:32(__init__)
   188751  148.820    0.001  213.559    0.001 enum.py:1226(__hash__)
   118324  139.084    0.001  176.961    0.001 ops.py:37(<listcomp>)
    33868  134.854    0.004 1487.571    0.044 ops.py:159(exec_ast)
    67346  130.721    0.002  286.326    0.004 helpers.py:153(__setitem__)
24320/20470  129.575    0.005  751.503    0.037 lazy.py:246(_push_movement_ops)
20216/16366  122.584    0.006 1383.653    0.085 lazy.py:263(elementwise_op)
    58632  117.923    0.002  178.214    0.003 tensor.py:16(__init__)
    61902  116.712    0.002  199.394    0.003 tensor.py:42(__init__)
    67346  102.056    0.002  216.656    0.003 helpers.py:140(remove)
   126386   96.639    0.001  137.998    0.001 helpers.py:132(__len__)
    76824   93.112    0.001  110.386    0.001 shapetracker.py:136(__init__)
    72466   90.195    0.001  396.054    0.005 helpers.py:154(__contains__)
    63190   88.715    0.001  179.039    0.003 helpers.py:127(_remove)
42946/37382   87.396    0.002  866.026    0.023 lazy.py:185(reshape)
94234/72648   86.658    0.001  297.978    0.004 ops.py:47(__hash__)
   210526   84.396    0.000   84.396    0.000 tensor.py:84(shape)
    68546   68.972    0.001   93.759    0.001 shapetracker.py:141(contiguous)
36804/10702   68.578    0.002  266.374    0.025 ops.py:52(map_buffers)
    19480   68.321    0.004 1023.489    0.053 tensor.py:230(reshape)
    71256   67.935    0.001   67.935    0.001 helpers.py:125(__init__)
     1940   67.530    0.035  311.900    0.161 tensor.py:265(__getitem__)
    67346   64.395    0.001   98.380    0.001 weakref.py:347(__new__)
    37779   61.867    0.002   90.833    0.002 lazy.py:293(__getitem__)
   169288   61.213    0.000   61.213    0.000 typing.py:2233(cast)
    28548   60.431    0.002  164.217    0.006 lazy.py:102(key)
    67346   57.226    0.001   57.226    0.001 weakref.py:352(__init__)
   147782   55.811    0.000   55.811    0.000 {built-in method builtins.isinstance}
    31638   55.109    0.002   55.109    0.002 shapetracker.py:36(__init__)
    67346   53.910    0.001   79.055    0.001 helpers.py:151(__len__)
    37302   53.896    0.001   84.441    0.002 enum.py:726(__contains__)
36804/10702   53.893    0.001  217.832    0.020 ops.py:52(<listcomp>)
   105185   50.281    0.000   50.281    0.000 shapetracker.py:144(shape)
23454/6608   50.159    0.002  110.558    0.017 ops.py:48(key)
     9129   49.137    0.005  161.446    0.018 symbolic.py:67(sum)
    94507   44.985    0.000   44.985    0.000 {built-in method builtins.max}
     2900   44.298    0.015 1557.607    0.537 tensor.py:470(dot)
76278/76256   42.367    0.001   48.988    0.001 {built-in method builtins.any}
     6608   42.247    0.006   70.566    0.011 ops.py:141(__call__)
    59647   41.204    0.001   41.204    0.001 helpers.py:47(value)
     4094   40.858    0.010  336.427    0.082 lazy.py:36(_ast_binaryops)
   106962   40.711    0.000   40.711    0.000 {method 'add' of 'set' objects}
```
This branch:

```
codegen         mean runtime:  137.03ms, runs:   157.13,  160.46,  121.59,  121.39,  125.52,  126.65,  173.97,  126.17,  132.45,  124.97
methodcache     mean runtime:  126.99ms, runs:   122.35,  164.48,  114.08,  121.26,  120.45,  119.81,  118.26,  116.95,  155.76,  116.52
profile         mean runtime:  695.42ms, runs:   742.53,  657.71,  689.83,  730.03,  662.94,  699.96,  689.63,  693.22,  695.22,  693.16
         5999385 function calls (5586548 primitive calls) in 6945.256 seconds

   Ordered by: internal time
   List reduced from 488 to 49 due to restriction <0.1>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
61170/2260  308.250    0.005 2228.884    0.986 lazy.py:109(realize)
    58632  277.281    0.005 3487.068    0.059 tensor.py:24(apply)
    66266  230.333    0.003  527.712    0.008 lazy.py:84(__init__)
    71376  200.954    0.003 1374.695    0.019 lazy.py:68(create_lazybuffer)
    43106  189.768    0.004 1327.206    0.031 lazy.py:169(shuffle_and_prune_movement_ops)
426150/359274  179.058    0.000  229.494    0.001 {built-in method builtins.len}
317423/206823  168.034    0.001  253.079    0.001 {built-in method builtins.hash}
    15086  166.139    0.011 2550.978    0.169 tensor.py:532(_broadcasted)
   205278  156.392    0.001  223.663    0.001 enum.py:1226(__hash__)
    41994  146.631    0.003 1290.634    0.031 ops.py:159(exec_ast)
   110571  140.201    0.001  304.216    0.003 ops.py:32(__init__)
24320/20470  136.452    0.006  714.228    0.035 lazy.py:248(_push_movement_ops)
   110571  129.454    0.001  164.015    0.001 ops.py:37(<listcomp>)
    66266  123.024    0.002  123.024    0.002 helpers.py:125(__init__)
20216/16366  121.716    0.006 1378.323    0.084 lazy.py:265(elementwise_op)
    62356  115.503    0.002  256.833    0.004 helpers.py:153(__setitem__)
    61902  111.054    0.002  197.899    0.003 tensor.py:42(__init__)
    58632  110.940    0.002  168.461    0.003 tensor.py:16(__init__)
   101950  102.727    0.001  140.276    0.001 helpers.py:133(add)
    62356   90.970    0.001  193.178    0.003 helpers.py:140(remove)
    71701   90.749    0.001  144.980    0.002 shapetracker.py:154(__init__)
   239033   88.797    0.000   88.797    0.000 {built-in method builtins.isinstance}
   119520   88.190    0.001  126.510    0.001 helpers.py:132(__len__)
    46061   83.861    0.002  113.666    0.002 shapetracker.py:306(movement_op)
   210526   82.002    0.000   82.002    0.000 tensor.py:84(shape)
    67476   81.326    0.001  357.440    0.005 helpers.py:154(__contains__)
89114/67628   79.014    0.001  268.074    0.004 ops.py:47(__hash__)
    45915   71.162    0.002  105.631    0.002 lazy.py:295(__getitem__)
    52644   70.719    0.001  143.267    0.003 helpers.py:127(_remove)
    19480   66.576    0.003  876.698    0.045 tensor.py:230(reshape)
     1940   66.256    0.034  292.814    0.151 tensor.py:265(__getitem__)
35002/11672   63.217    0.002  250.120    0.021 ops.py:52(map_buffers)
    30488   62.803    0.002  132.224    0.004 lazy.py:102(key)
    45316   62.407    0.001   97.662    0.002 enum.py:726(__contains__)
    62356   58.273    0.001   88.830    0.001 weakref.py:347(__new__)
    62356   52.500    0.001   52.500    0.001 weakref.py:352(__init__)
    51216   51.619    0.001   70.114    0.001 shapetracker.py:159(contiguous)
35002/11672   49.706    0.001  198.079    0.017 ops.py:52(<listcomp>)
    62356   47.850    0.001   70.478    0.001 helpers.py:151(__len__)
21652/7578   47.004    0.002  101.971    0.013 ops.py:48(key)
     7578   45.808    0.006   77.134    0.010 ops.py:141(__call__)
     7578   43.974    0.006  476.344    0.063 linearizer.py:93(__init__)
    70738   43.807    0.001   43.807    0.001 helpers.py:47(value)
    94506   43.761    0.000   43.761    0.000 {built-in method builtins.max}
    89361   42.895    0.000   42.895    0.000 shapetracker.py:162(shape)
    45914   42.374    0.001   69.794    0.002 helpers.py:43(__bool__)
   115571   41.121    0.000   41.121    0.000 typing.py:2233(cast)
37688/35718   39.757    0.001  577.591    0.016 lazy.py:187(reshape)
     2900   39.022    0.013 1311.900    0.452 tensor.py:470(dot)
```

**Comment by geohot** [comment]
How does this handle multiview shapetrackers?

**Comment by roelofvandijk** [comment]
This is a pure rewrite of the functions that were already there, reducing the number of arguments that get passed to each function to the necessary ones (and no `self`).

**Comment by geohot** [comment]
I see, yea I'm generally supportive of this approach, I think I understand it as "immutable shapetrackers"

**Pull request closed without merging**